core/shared/src/main/scala/fs2/concurrent/TimedPull.scala

```diff
@@ -0,0 +1,332 @@
+
+        // TODO is the initial time.get.unNone fine or does it spin?
+        def timeouts: Stream[F, Token] =
+          Stream.eval(time.get).unNone.flatMap { timeout =>
+            Stream.eval(timeout.asOfNow).flatMap { t =>
+              if (t <= 0.nanos) Stream.emit(timeout.id) ++ nextAfter(timeout).drain
+              else Stream.sleep_[F](t)
+            }
+          } ++ timeouts
+
+        def output: Stream[F, Either[Token, Chunk[A]]] =
+          timeouts
+            .map(_.asLeft)
+            .mergeHaltR(source.chunks.map(_.asRight))
+            .evalMapFilter { // TODO filter of stale timeouts in uncons
+
+              case Right(c) => c.asRight[Token].some.pure[F]
+              case Left(id) => time.get.map(t => t.filter(_.id != id).as(id.asLeft[Chunk[A]]))
+            }
+
+        def toTimedPull(s: Stream[F, Either[Token, Chunk[A]]]): TimedPull[F, A] = new TimedPull[F, A] {
+          type Timeout = Token
+
+          def uncons: Pull[F, INothing, Option[(Either[Token, Chunk[A]], TimedPull[F, A])]] =
+            s.pull.uncons1
+              .map(_.map(_.map(toTimedPull)))
+
+          def startTimer(t: FiniteDuration): Pull[F, INothing, Unit] = Pull.eval {
+            newTimeout(t).flatMap(t => time.set(t.some))
+          }
+        }
+
+        pull(toTimedPull(output)).stream
+      }
+    }
+  }
+
+  // TODO both old and new impl probably have a corner case with timing out on the first chunk
+  // think about that
+  // Actually it doesn't affect only the first chunk, but every time there has been a timeout
+  // but no chunks currently emitted (empty acc)
+
+  // so let's say we have groupWithin(10, 1.second). A stream emits 10
+  // elems, so groupWithin emits those and restarts go, with an empty
+  // Chunk queue. Then the stream sleeps 1 second, groupWithin hits the
+  // timeout, emits nothing, and resets the timer. The stream sleep 200
+  // mills, and emits 3 elems. In the current implementation, that
+  // something is not emitted, because the timer has been resetted (and
+  // is not at 800 mills). But from the persective of the returned
+  // Stream, we have been waiting for an element for 1.2 seconds, and
+  // even though 3 elems have been emitted by the original stream, those
+  // are not emitted in the result
+  //
+  //  A simpler case of this problem would be a timeout before the first
+  // chunk of the stream arrives: again say the stream sleeps 1.2 seconds
+  // on then emits 3 elems, the current implementation will wait another
+  // 800 millis before emitting those 3, whereas arguably they should be
+  // emitted immediately
+  //
+  // a solution would be to go in another recursive state, has timed out,
+  // which emits the next chunk as soon as it arrives, and resets the timer then.
+  // the question is what to do if the chunk which arrives after the limit is bigger
+  // than the size: do we emit all of it (after splitting), or do we accumulate the remainder
+  // for the next chunk?
+
```

Comment:
I'd like to get some opinions here. I think my preferred semantics is: if a timeout triggers but there are accumulated elements, then the timer is not reset. When the next chunk arrives, up to `n` elements of it are emitted immediately, the timer is reset, and the remainder (if any) is carried in the next iteration of `go`.

Reply:
The main point of doubt for me is between this

> up to n elements of it are emitted immediately [.. ]and the remainder (if any) is carried in the next iteration of `go`

and emitting the remainder immediately as well

Reply:
I have settled for the semantics described above